src/attr.rs

```diff
@@ -195,15 +195,20 @@ impl<'s> Attributes<'s> {
         self.0.as_ref().map_or(true, |v| v.is_empty())
     }
 
+    #[must_use]
+    pub fn len(&self) -> usize {
+        self.0.as_ref().map_or(0, |v| v.len())
+    }
+
     /// Returns a reference to the value corresponding to the attribute key.
     #[must_use]
-    pub fn get(&self, key: &str) -> Option<&AttributeValue<'s>> {
+    pub fn get(&'s self, key: &str) -> Option<&'s AttributeValue<'s>> {
         self.iter().find(|(k, _)| *k == key).map(|(_, v)| v)
     }
 
     /// Returns an iterator over the attributes in undefined order.
-    pub fn iter(&self) -> impl Iterator<Item = (&'s str, &AttributeValue<'s>)> + '_ {
-        self.0.iter().flat_map(|v| v.iter().map(|(a, b)| (*a, b)))
+    pub fn iter(&'s self) -> AttrIter<'s> {
```

Comment:
This one does need to stay, unfortunately, again because of the parameterisation of the iterator, I think.

Reply:
again, should be possible without <'s> on AttributeIter?

Reply:
Yes, absolutely correct - but iter_mut does need them (I've been wrong before but I did actually try it this time...)

Reply:
hmm, I think this may cause problems when using the functions. might be good to have a simple test case that just tries modifying some value and checking the result and make sure it compiles without any lifetime issues.

I think we need two lifetimes: one for the immutable source code ('s), and one for the mutable reference to the Attributes vector. I.e. I think the AttributesIterMut may need to look something like:

```rust
pub struct AttributesIterMut<'a, 's>(std::slice::IterMut<'a, (&'s str, AttributeValue<'s>)>)
```

Might be a good idea to do the same for AttributesIter to make it explicit that we have two separate lifetimes.